core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -245,6 +257,28 @@ private[spark] class TaskSetManager(
     allPendingTasks += index  // No point scanning this whole list to find the old task there
   }
 
+  private[spark] def addPendingSpeculativeTask(index: Int) {
+    for (loc <- tasks(index).preferredLocations) {
+      loc match {
+        case e: ExecutorCacheTaskLocation =>
+          pendingSpeculatableTasksForExecutor.getOrElseUpdate(
+            e.executorId, new ArrayBuffer) += index
+        case _ =>
+      }
+      pendingSpeculatableTasksForHost.getOrElseUpdate(loc.host, new ArrayBuffer) += index
+      for (rack <- sched.getRackForHost(loc.host)) {
+        pendingSpeculatableTasksForRack.getOrElseUpdate(rack, new ArrayBuffer) += index
+      }
+    }
+
+    if (tasks(index).preferredLocations == Nil) {
+        pendingSpeculatableTasksWithNoPrefs += index
+    }
+
+    // No point scanning this whole list to find the old task there
+    allPendingSpeculatableTasks += index
```

Comment:
qq does this need to hold all tasks or only with TaskLocality.ANY pref... seems like a waste to scan all the tasks within it for getting non local tasks, could be wrong in my understanding however thanks

Reply:
Also in your tests would be interesting to note what is locality affinity... do you see more process local tasks vs node local etc., that might be related to scheduling logic, just curious to know about it.

Reply:
So afaik, task locality here means a preference for the individual task to run at a desired locality level(kind of like for flight scheduling, my preference would be to fly from airport A as it is the nearest to my house, but if the flight i want to book does not take off from location A, location B which is farther than A, will do). If the desired locality level cannot be assigned to the task by the scheduler due to locality constraints and other checks, it will be assigned the next best locality. That is why we need all tasks in the ANY locality preference queue. The method `hasAttemptOnHost` will take care of duplicate scheduling of tasks anyways.

Reply:
Good point. Will definitely test for locality affinity, but ideally, that logic should not have changed with this PR.